File: studentassistant/src/main/java/ru/erdenian/studentassistant/schedule/Homework.java
package ru.erdenian.studentassistant.schedule;

import org.joda.time.LocalDate;
import org.joda.time.LocalTime;

import lombok.NonNull;
import lombok.Value;

/**
 * Класс домашнего задания. Является неизменяемым.
 *
 * @author Ilya Solovyev
 * @version 0.0.0
 * @since 0.0.0
 */
@Value
public class Homework implements Comparable<Homework> {

    /**
     * Идентификатор домашнего задания. Должен быть уникальным для каждого задания. При
     * редактировании существующего домашнего задания должен создаваться новый объект с таким
     * же id, старый объект должен быть удален.
     *
     * @since 0.0.0
     */
    private final long id;

    /**
     * Название предмета.
     *
     * @since 0.0.0
     */
    @NonNull
    private final String subjectName;

    /**
     * Описание.
     *
     * @since 0.0.0
     */
    @NonNull
    private final String description;

    /**
     * День сдачи.
     *
     * @since 0.0.0
     */
    @NonNull
    private final LocalDate deadlineDay;

    /**
     * Время сдачи.
     *
     * @since 0.0.0
     */
    @NonNull
    private final LocalTime deadlineTime;

    /**
     * В качестве идентификатора домашнего задания используется {@link System#nanoTime()}.
     *
     * @param subjectName  название предмета ({@link Homework#subjectName})
     * @param description  описаниеа ({@link Homework#description})
     * @param deadlineDay  день сдачи ({@link Homework#deadlineDay})
     * @param deadlineTime время сдачи ({@link Homework#deadlineTime});
     *                     если передать null, запишется 23:59:59
     * @since 0.0.0
     */
    public Homework(@NonNull String subjectName, @NonNull String description,
                    @NonNull LocalDate deadlineDay, @NonNull LocalTime deadlineTime) {

        this(System.nanoTime(), subjectName, description, deadlineDay, deadlineTime);
    }

    /**
     * @param id           идентификатор домашнего задания ({@link Homework#id})
     * @param subjectName  название предмета ({@link Homework#subjectName})
     * @param description  описание ({@link Homework#description})
     * @param deadlineDay  день сдачи ({@link Homework#deadlineDay})
     * @param deadlineTime время сдачи ({@link Homework#deadlineTime});
     *                     если передать null, запишется 23:59:59
     * @since 0.0.0
     */
    public Homework(long id, @NonNull String subjectName, @NonNull String description,
                    @NonNull LocalDate deadlineDay, LocalTime deadlineTime) {

        this.id = id;
        this.subjectName = subjectName;
        this.description = description;
        this.deadlineDay = deadlineDay;
        this.deadlineTime = deadlineTime != null ? deadlineTime : new LocalTime(23, 59, 59);
    }

    /**
     * @since 0.0.0
     */
    @Override
    public int compareTo(@android.support.annotation.NonNull Homework homework) {
        return 0;
    }
}

File: studentassistant/src/main/kotlin/ru/erdenian/studentassistant/schedule/Homework.kt
package ru.erdenian.studentassistant.schedule

import com.google.common.collect.ComparisonChain
import org.joda.time.LocalDate

/**
 * Класс домашнего задания.
 *
 * @author Ilya Solovyev
 * @since 0.0.0
 * @property subjectName название предмета, по которому задано занание
 * @property description описание задания
 * @property deadline срок сдачи
 * @property id уникальный id задания
 * @throws IllegalArgumentException если [subjectName] или [description] пусты
 */
data class Homework(val subjectName: String, val description: String, val deadline: LocalDate,
                    val id: Long = generateId()) : Comparable<Homework> {

  init {
    if (subjectName.isBlank()) throw IllegalArgumentException("Пустое название предмета")
    if (description.isBlank()) throw IllegalArgumentException("Пустое описание")
  }

  override fun compareTo(other: Homework) = ComparisonChain.start()
      .compare(deadline, other.deadline)
      .compare(description, other.description)
      .compare(id, other.id)
      .result()
}
